G1ANT.Addon.Net/SimplifiedMessageSummary.cs

```diff
@@ -1,27 +1,321 @@
+                return null;
+            }
+        }
+
+        public MailboxAddress Sender
+        {
+            set
+            {
+                if (fullMessage != null)
+                    fullMessage.Sender = value;
+                else
+                    throw new NotSupportedException("Cannot set Sender for recieved message");
+            }
+        }
+
+        public InternetAddressList Cc
+        {
+            get
+            {
+                if (fullMessage != null)
+                    return fullMessage.Cc;
+                else if (messageSummary != null)
+                    return messageSummary.Envelope.Cc;
+                return null;
+            }
+        }
+
+        public InternetAddressList Bcc
+        {
+            get
+            {
+                if (fullMessage != null)
+                    return fullMessage.Bcc;
+                else if (messageSummary != null)
+                    return messageSummary.Envelope.Bcc;
+                return null;
+            }
+        }
+
+        public bool IsReply
+        {
+            get
+            {
+                if (messageSummary != null)
+                    return messageSummary.IsReply;
+                return false;
+            }
+        }
+
+        public string Priority
+        {
+            get
+            {
+                if (fullMessage != null && fullMessage.Headers != null)
+                    return fullMessage.Headers[HeaderId.Priority];
+                else if (messageSummary != null && messageSummary.Headers != null)
+                    return messageSummary.Headers[HeaderId.Priority];
+                return "0";
+            }
+            set
+            {
+                if (fullMessage != null && fullMessage.Headers != null)
+                    fullMessage.Headers[HeaderId.Priority] = value;
+                else if (messageSummary != null && messageSummary.Headers != null)
+                    messageSummary.Headers[HeaderId.Priority] = value;
+            }
+        }
+
+        public DateTimeOffset? Date
+        {
+            get
+            {
+                if (fullMessage != null)
+                    return fullMessage.Date;
+                else if (messageSummary != null)
+                    return messageSummary.Envelope.Date;
+                return null;
+            }
+            set
+            {
+                if (fullMessage != null && value.HasValue)
+                    fullMessage.Date = value.Value;
+                else if (messageSummary != null)
+                    messageSummary.Envelope.Date = value;
+            }
+        }
+
+        private void FillBodyBuilderAttachments(BodyBuilder bodyBuilder)
+        {
+            if (bodyBuilder.Attachments != null)
+                foreach (var att in FullMessage.Attachments)
+                    bodyBuilder.Attachments.Add(att);
+        }
+
+        public string HtmlBody
+        {
+            get => FullMessage.HtmlBody;
```

Comment:
Are we still in the same model, where once `fullMessage` is used and other time `messageSummary`?

Also, take a look: if you call `FullMessage` when `fullMessage` is null and `messageSummary` is not then you end up with both `fullMessage != null` and `messageSummary != null`. But a moment ago you initialized this class with `messageSummary` and `fullMessage` should be null. Now all getters return fields from different object.
What caused that was a simple call to `HtmlBody`. Come on, why does `HtmlBody` cause switch to another data source (summary=>full)? It's not called `GetHtmlBodyAndSwitchDataSourceToFullMessage`.

I read the code of whole class once again and I finally got your idea, but I'm still sure that this class should become two separate models + 3rd model that joins them together to achieve seamless/transparent conversion of data source.

Reply:
If we want to make it right we should also rewrite smtp commands which retrieve emails. In general it's to much time to spent to refactor and rewrite everything